File: solution1.py
# https://leetcode.com/problems/paint-house

# // Time Complexity : o(n*3) n= number of houses
# // Space Complexity : o(n*3) n= number of houses
# // Did this code successfully run on Leetcode :yes
# // Any problem you faced while coding this : None
#
#
# // Your code here along with comments explaining your approach

class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        dp = [[0 for i in range(3)] for j in range(len(costs) + 1)]
        for i in range(1, len(costs) + 1):
            for j in range(3):
                # red
                if j == 0:
                    dp[i][j] = costs[i - 1][j] + min(dp[i - 1][1], dp[i - 1][2])

                # blue
                if j == 1:
                    dp[i][j] = costs[i - 1][j] + min(dp[i - 1][0], dp[i - 1][2])

                # green
                if j == 2:
                    dp[i][j] = costs[i - 1][j] + min(dp[i - 1][0], dp[i - 1][1])

        return min(dp[len(costs)][0], dp[len(costs)][1], dp[len(costs)][2])

File: Solution1.java
// Leetcode Problem : https://leetcode.com/problems/paint-house/
// Time Complexity : O(3n) where n is the number of houses
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*
 * I start iterating from the second row of the costs matrix. For the element in first column,
 * I replace the value as the sum of cost at that position + minimum of cost at previous row
 * but second and third columns(so that colors aren't repeated). Similarly, 
 * costs[i][1]+=Math.min(costs[i-1][2],costs[i-1][0]) and costs[i][2]+=Math.min(costs[i-1][1],costs[i-1][0]);
 */

class Solution {
    public int minCost(int[][] costs) {
        for(int i=1;i<costs.length;i++)
        {
            
            costs[i][0]+=Math.min(costs[i-1][1],costs[i-1][2]);
            costs[i][1]+=Math.min(costs[i-1][2],costs[i-1][0]);
            costs[i][2]+=Math.min(costs[i-1][1],costs[i-1][0]);
            
        }
        return Math.min(Math.min(costs[costs.length-1][0],costs[costs.length-1][1]),costs[costs.length-1][2]);
    }
}